include/unifex/upon_done.hpp

```diff
@@ -0,0 +1,169 @@
+#pragma once
+
+#include <unifex/config.hpp>
+#include <unifex/async_trace.hpp>
+#include <unifex/bind_back.hpp>
+#include <unifex/receiver_concepts.hpp>
+#include <unifex/sender_concepts.hpp>
+#include <unifex/std_concepts.hpp>
+#include <unifex/tag_invoke.hpp>
+#include <unifex/type_list.hpp>
+#include <unifex/type_traits.hpp>
+
+#include <functional>
+
+#include <unifex/detail/prologue.hpp>
+
+namespace unifex {
+namespace _upon_done {
+
+template <typename Receiver, std::invocable Func>
```

Comment:
Libunifex targets C++17 compilers that don't support concepts. And generally we don't constrain implementation details in the interest of compile times. So check parameters at interface boundaries, and constrain overloads like `connect`, `set_*`, and `start`, and otherwise use unconstrained templates.